pkgs/fake_async/lib/fake_async.dart

```diff
@@ -207,23 +210,25 @@ class FakeAsync {
   /// The [timeout] controls how much fake time may elapse before a [StateError]
   /// is thrown. This ensures that a periodic timer doesn't cause this method to
   /// deadlock. It defaults to one hour.
-  void flushTimers(
-      {Duration timeout = const Duration(hours: 1),
-      bool flushPeriodicTimers = true}) {
+  void flushTimers({
+    Duration timeout = const Duration(hours: 1),
+    bool flushPeriodicTimers = true,
+  }) {
     final absoluteTimeout = _elapsed + timeout;
     _fireTimersWhile((timer) {
       if (timer._nextCall > absoluteTimeout) {
         // TODO(nweiz): Make this a [TimeoutException].
         throw StateError('Exceeded timeout $timeout while flushing timers');
       }
 
-      if (flushPeriodicTimers) return _timers.isNotEmpty;
+      // Always run timer if it's due.
+      if (timer._nextCall <= elapsed) return true;
 
-      // Continue firing timers until the only ones left are periodic *and*
-      // every periodic timer has had a change to run against the final
-      // value of [_elapsed].
-      return _timers
-          .any((timer) => !timer.isPeriodic || timer._nextCall <= _elapsed);
+      // If no timers are due, continue running timers
+      // (and advancing time to their next due time)
+      // if flushing periodic timers,
+      // or if there is any non-periodic timer left.
+      return flushPeriodicTimers || _timers.any((timer) => !timer.isPeriodic);
```

Comment:
This function should just check whether to run the *current* timer, which it should unless it's not due, we're not flushing periodic timers, and all remaining timers (including this one) are periodic timers.